zio-http/shared/src/main/scala/zio/http/codec/HttpContentCodec.scala

```diff
@@ -0,0 +1,195 @@
+package zio.http.codec
+
+import scala.collection.immutable.ListMap
+
+import zio._
+
+import zio.stream.ZPipeline
+
+import zio.schema.Schema
+import zio.schema.codec._
+
+import zio.http._
+import zio.http.codec.HttpContentCodec.json.fromSchema
+
+final case class HttpContentCodec[A](
+  choices: ListMap[MediaType, BinaryCodec[A]],
+  schema: Schema[A],
+) { self =>
+
+  /**
+   * A right biased merge of two HttpContentCodecs.
+   */
+  def ++(that: HttpContentCodec[A]): HttpContentCodec[A] =
+    HttpContentCodec(choices ++ that.choices, schema)
+
+  def decodeRequest(request: Request): Task[A]    = {
+    val contentType =
+      request.header(Header.ContentType).getOrElse(Header.ContentType(MediaType.application.`json`))
+    lookup(contentType) match {
+      case Some(codec) =>
+        request.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+  def decodeResponse(response: Response): Task[A] = {
+    val contentType =
+      response.header(Header.ContentType).getOrElse(Header.ContentType(MediaType.application.`json`))
+    lookup(contentType) match {
+      case Some(codec) =>
+        response.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+
+  def encode(value: A): Either[String, Body] =
```

Comment:
This is hot path, too. Look at allocations:

1. headOption (1)
2. Tuple2 (1)
3. Body (2)
4. Either (1)

Some are unavoidable, but some can be avoided. For example, on `choices` use `nonEmpty` and `head`. Do not pattern match on the tuple, rather extract out the components individually. And construct `Body` with both the byte **and** the media type at the same time.